File: comparisons/lmm_pastes.jl
# Please install the required packages before running:
# import Pkg; Pkg.add(["CSV", "DataFrames", "MixedModels"])

using CSV
using DataFrames
using MixedModels

function main()
    file_path = joinpath("tests", "data", "pastes.csv")
    if !isfile(file_path)
        println("Could not find the dataset at: $file_path")
        println("Please run this example from the root of the lme-rs repository.")
        exit(1)
    end

    println("Loading data from $file_path...")
    
    # 1. Load the dataset
    df = CSV.read(file_path, DataFrame)

    # 2. Fit the Linear Mixed Model
    println("\nFitting model: strength ~ 1 + (1 | batch) + (1 | batch:cask)")
    
    # MixedModels.jl evaluates nested effectively as batch + batch:cask or uses zerocorr(batch&cask)
    form = @formula(strength ~ 1 + (1 | batch) + (1 | batch&cask))
    
    m1 = fit(MixedModel, form, df, REML=true)
    
    # 3. Print the summary
    println("\n=== Model Summary ===")
    println(m1)

    # 4. Generate Predictions
    println("\n=== Predictions ===")
    println("Generating predictions for population-level...")
    
    # Predict (Population-level)
    # \mu = X \beta
    beta = coef(m1)
    X = ones(3, 1) # Intercept only for fixed effects
    
    preds = X * beta
    
    println("Predictions (Population-level):")
    println(preds)
end

main()

File: comparisons/lmm_pastes.py
"""
Pastes LMM using statsmodels.

`lme4::lmer(strength ~ 1 + (1 | batch/cask))` uses two variance components
(batch and batch:cask). `statsmodels` `MixedLM` supports only one grouping
column per fit. We group by the `sample` column (`batch:cask` labels), which
matches the inner random intercept stratum only — not the full nested model.
For exact nested parity, use R, Julia, or Rust examples in this folder.
"""

from __future__ import annotations

import os
import sys

try:
    import pandas as pd
    import statsmodels.formula.api as smf
except ImportError:
    print("Please install pandas and statsmodels:")
    print("pip install pandas statsmodels")
    sys.exit(1)


def main() -> None:
    path = os.path.join("tests", "data", "pastes.csv")
    if not os.path.exists(path):
        print(f"Could not find the dataset at {path}")
        print("Please run this example from the root of the lme-rs repository.")
        sys.exit(1)

    data = pd.read_csv(path)
    print("\nFitting model: strength ~ 1 + (1 | sample)  [sample = batch:cask level]")
    print("(statsmodels single-grouping proxy for nested batch/cask)")

    model = smf.mixedlm("strength ~ 1", data, groups=data["sample"])
    result = model.fit(reml=True)
    print("\n=== Model Summary ===")
    print(result.summary())

    print("\n=== Predictions (population-level) ===")
    fe = result.fe_params
    print("Intercept (population mean):", float(fe["Intercept"]))


if __name__ == "__main__":
    main()
